sphinx_gallery/recommender.py

```diff
@@ -0,0 +1,273 @@
+        if not isinstance(n_examples, numbers.Integral):
+            raise ValueError("n_examples must be an integer")
+        elif n_examples < 1:
+            raise ValueError("n_examples must be strictly positive")
+
+        freq_func = self.token_freqs
+        counts_matrix = self.dict_vectorizer(
+            [freq_func(Path(fname).read_text()) for fname in file_names]
+        )
+
+        if isinstance(min_df, float) and 0.0 < min_df <= 1.0:
+            min_df = int(np.ceil(min_df * counts_matrix.shape[0]))
+        elif not isinstance(min_df, numbers.Integral) and min_df > 1:
+            raise ValueError("min_df must be float in range [0.0, 1.0] or int")
+        if isinstance(max_df, float) and 0.0 < max_df <= 1.0:
+            max_df = int(np.floor(max_df * counts_matrix.shape[0]))
+        elif not isinstance(max_df, numbers.Integral) and max_df > 1:
+            raise ValueError("max_df must be float in range [0.0, 1.0] or int")
+        doc_appearances = np.sum(counts_matrix, axis=0)
+        mask = (doc_appearances >= min_df) & (doc_appearances <= max_df)
+        self.similarity_matrix_ = self.cosine_similarity(
+            self.compute_tf_idf(counts_matrix[:, mask])
+        )
+        self.file_names_ = file_names
+        return self
+
+    def predict(self, file_name):
+        """Compute the most `n_examples` similar documents to the query.
+
+        Parameters
+        ----------
+        file_name : str
+            Name of the file corresponding to the query index `item_id`.
+
+        Returns
+        -------
+        recommendations : list of str
+            Name of the files most similar to the query.
+        """
+        item_id = self.file_names_.index(file_name)
+        similar_items = list(enumerate(self.similarity_matrix_[item_id]))
+        sorted_items = sorted(similar_items, key=lambda x: x[1], reverse=True)
+
+        # Get the top k items similar to item_id. Note that `sorted_items[0]`
+        # is always the query document itself, hence it is discarded from the
+        # returned list of recommendations.
+        top_k_items = [idx for idx, _ in sorted_items[1 : self.n_examples + 1]]
+        recommendations = [self.file_names_[idx] for idx in top_k_items]
+        return recommendations
+
+
+def _write_recommendations(recommender, fname, gallery_conf):
+    """Generate `.recommendations` RST file for a given example.
+
+    Parameters
+    ----------
+    recommender : ExampleRecommender
+        Instance of a fitted ExampleRecommender.
+
+    fname : str
+        Path to the example file.
+
+    gallery_conf : dict
+        Configuration dictionary for the sphinx-gallery extension.
+    """
+    path_fname = Path(fname)
+    recommend_fname = f"{path_fname.parent / path_fname.stem}.recommendations.new"
+    recommended_examples = recommender.predict(fname)
+
+    with open(recommend_fname, "w", encoding="utf-8") as ex_file:
+        ex_file.write("\n\n.. rubric:: Related examples\n")
```

Comment:
Based on @betatim's suggestion in the linked scikit-learn PR.

```suggestion
        ex_file.write(
            "\n\n.. rubric:: Automatically generated list of related examples\n"
        )
```

Maybe you should make this rubric header user-settable via a configuration parameter instead of hard-coding it in the Python code.

```suggestion
     default_rubric_header = (
         ".. rubric:: Automatically generated list of related examples"
     )
     rubric_header = gallery_conf['recommender'].get(
         'rubric_header', default_rubric_header
     )
     ex_file.write(f"\n\n{rubric_header}\n")
```

This way it gives full freedom to customize this later (e.g. add a tooltip or a small paragraph to explain how the examples are generated).